### DISCUSSION TITLE
Connectivity_matrix is (too ?) slow

### DISCUSSION THREAD

**Discussion by DelinteNicolas** [dipy/dipy#2972] [Tractography / Tractometry]
As mentioned in the function description, `connectivity_matrix` is slow when 'inclusive' is set to True. I believe that a faster function could help Dipy users, so here is a rough code that runs faster when the 'inclusive' parameter is set to True.
I tried to add the other parameters present in the Dipy function, but these were not tested extensively. The affine, set_coordinates,... lines are also absent since they were not needed in my case and I am not familiar with the terminology, so I opted for a simpler version of the code rather than to add lines where I could easily make a mistake.
Nonetheless, the code with 'inclusive' parameter now runs under 10 seconds for a tractogram of 500 000 streamlines, compared to around a minute with the old code.

New code template:

    def connectivity_matrix(streamlines, label_volume, inclusive: bool = True, symmetric: bool = True, return_mapping: bool = False, return_mapping_as_streamlines: bool = False):

      label_volume = label_volume.astype(int)
      matrix = np.zeros((np.max(label_volume)+1, np.max(label_volume)+1))
  
      mapping = defaultdict(list)
  
      if inclusive:
          for i, sl in enumerate(streamlines):
  
              x, y, z = np.floor(sl.T).astype(int)
              crossed_labels = np.unique(label_volume[x, y, z])
  
              for comb in combinations(crossed_labels, 2):
                  matrix[comb] += 1
  
                  if return_mapping:
                      mapping[comb].append(i)
                  if return_mapping_as_streamlines:
                      mapping[comb].append(sl)
  
      else:
          streamlines = np.array([sl[0::len(sl)-1] for sl in streamlines])
          x, y, z = np.floor(streamlines.T).astype(int)
          end_labels = np.sort(label_volume[x, y, z], axis=0)
          np.add.at(matrix, (end_labels[0].T, end_labels[1].T), 1)
  
          if return_mapping:
              for i, (a, b) in enumerate(end_labels.T):
                  mapping[a, b].append(i)
  
          if return_mapping_as_streamlines:
              for i, (a, b) in enumerate(end_labels.T):
                  mapping[a, b].append(streamlines[i])
  
      if symmetric:
          matrix = matrix+matrix.T
  
      if return_mapping or return_mapping_as_streamlines:
          return (matrix, mapping)
      else:
          return matrix`

Hope this helps

**Comment by skoudoro** [comment]
Thank you for this @DelinteNicolas .

I would be great if you could create a PR Directly. will you feel comfortable to do it?

Thank you for your feedback

**Reply by DelinteNicolas** [reply]
Hello, I have made a pull request (#2975 ). I tried to add the 'affine' lines and parameters.
